### DISCUSSION TITLE
Is it possible to have Error/Catch Boundaries on individual components (not routes)?

### DISCUSSION THREAD

**Discussion by joelrozen** [remix-run/remix#4971] [Q&A]
Apologies if this is obvious but I've scoured through the docs and also watched Full Stack Components vid and read through the article - but let's say I have a dashboard with n number of widgets. Each widget needs its own data from the database. From what I can gather, I can 
A) fetch EVERYTHING I need in the route loader and then pass each individual bit down as props to each widget; or
B) each component makes its own call to get the info it needs

A has the disadvantage of if any single thing goes wrong, the whole thing goes wrong, while B has the disadvantage of the waterfall networking effect.

So I have tried to use Error and Catch Boundaries in my components (note, NOT routes) like so (minimal example):
```typescript
// /routes/dashboard/index.tsx

export default function DashboardHome() {
  return (
    <LayoutStuff>
      <Widget1 />
      <Widget2 />
      <Widget3 />
      <Widget4 />
      ...
      <WidgetN />
    </LayoutStuff>
  );
}
```

```typescript
// /routes/resources/widgets/widget1.tsx (as per FSC recommendation)

export const loader = async () => {
  return json({ widget_1_info: 25 });
};

export const Widget1 = () => {
  const fetcher = useFetcher<typeof loader>();

  useEffect(() => {
    if (!fetcher.data) {
      fetcher.load("/resources/widgets/widget1");
    }
  });

  if (!fetcher.data) return <Spinner />;

  return <pre>{JSON.stringify(fetcher.data, null, 2)}</pre>;
};

export const CatchBoundary = () => {
  return <Text colorScheme={"red"}>Something went wrong</Text>;
};

export const ErrorBoundary = () => {
  return <Text colorScheme={"red"}>Something went wrong</Text>;
};
```

Now when the data works, the spinner displays, then it JSON prints the response. Awesome. But if I throw a Response from the loader here, or something else goes wrong, I want *only* that widget to display its boundary, that way the other N-1 widgets still display. As it stands, because Widget1 isn't a "route" component, the Error/Catch Boundaries are bubbling up to the Route, and then the entire dashboard renders the Boundary.

Is there a way to render the boundary on the individual component level? Even if I fetched all the data *seperately*, if any of the individual routes the components fetch throws, then it still keeps going up. 

What is the best way to approach this situation without having to fetch *everything* in the parent loader function (since end result is the same - everything displays or nothing displays).

Note: possibly related to [4116](https://github.com/remix-run/remix/discussions/4116) - seems similar-ish?

**Comment by lili21** [comment]
You can, but not with `fetcher.load`. You can use some data fetching library which support suspense, like react-query or swr

**Comment by lili21** [comment]
The `CatchBoundary` and `ErrorBoundary` defined in route module only work for route level. 
If you want the component level boundary, You need to create the `ErrorBoundary` component  - [docs](https://reactjs.org/docs/error-boundaries.html)

**Reply by joelrozen** [reply]
This still doesn't work though as the loader is throwing a response // the "error" happens server side, and even if my component with the useFetcher is wrapped in an ErrorBoundary, it still goes up to the route ErrorBoundary.

**Reply by lili21** [reply]
You can't use fetcher.load. You need to use a data fetching library which support suspense fetch the data.

**Reply by lili21** [reply]
if you want fetch it on the server side, then use the `defer` API. 
if you want fetch it on the client side, use react-query or swr with suspense.

**Reply by joelrozen** [reply]
Gotcha. Thanks!

**Comment by lili21** [comment]
I guess the best way is using `defer`.  You can fetching everything in the loader, which prevent the waterfall issue. But you don't need to wait the fetching.

**Reply by lili21** [reply]
the `defer` API is on developing. if you want try it right now, you can install the `deferred` version of Remix, and here is the [guide](https://github.com/remix-run/remix/blob/deferred/docs/guides/streaming.md)

**Reply by joelrozen** [reply]
I will give it a go and report back.
